### PR TITLE
net: lwm2m: add callback for send confirmation

### PR DESCRIPTION
**Pull request by RomainPelletant**
Make send result deprecated as confirmation result shall always be asked. Add user callback to get confirmation result. See #52328

Signed-off-by: romain pelletant <romainp@kickmaker.net>

### CODE REVIEW

**Comment by RomainPelletant** [comment]
Add @broglep-work , @SeppoTakalo

File: include/zephyr/net/lwm2m.h [hunk 1]
```diff
@@ -1410,7 +1410,23 @@ void lwm2m_rd_client_update(void);
 char *lwm2m_path_log_buf(char *buf, struct lwm2m_obj_path *path);
 
 /** 
- * @brief LwM2M SEND operation to given path list
+ * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
+ *
+ * @param ctx LwM2M context
+ * @param path_list LwM2M Path string list
+ * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
+ * @param reply_cb Callback triggered with confirmation state.
```
Comment (SeppoTakalo) [inline include/zephyr/net/lwm2m.h:1418]: I think the callback should be optional, so this parameter should allow NULL.

File: include/zephyr/net/lwm2m.h [hunk 2]
```diff
@@ -1410,7 +1410,23 @@ void lwm2m_rd_client_update(void);
...
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_engine_send_cb(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
+		     coap_reply_t *reply_cb);
+
+/** 
+ * @brief LwM2M SEND operation to given path list asynchronously without confirmation
+ *
+ * @deprecated Use lwm2m_engine_send_cb() instead, to check for confirmation request callbacks
```
Comment (SeppoTakalo) [inline include/zephyr/net/lwm2m.h:1429]: Use also `__deprecated` before the function so compiler can warn about it.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 3]
```diff
@@ -2915,26 +2915,6 @@ int do_composite_read_op_for_parsed_list(struct lwm2m_message *msg, uint16_t con
...
-		LOG_INF("Send done!");
-		return 0;
-	}
-
-	LOG_ERR("Failed with code %u.%u. Not Retrying.", COAP_RESPONSE_CODE_CLASS(code),
-		COAP_RESPONSE_CODE_DETAIL(code));
-
-	return 0;
-}
-
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:2937]: Leave this callback as a default callback.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 4]
```diff
@@ -3055,15 +3035,15 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
...
-		msg->reply_cb = do_send_reply_cb;
-		msg->message_timeout_cb = do_send_timeout_cb;
-	} else {
-		msg->type = COAP_TYPE_NON_CON;
-		msg->reply_cb = NULL;
-		msg->message_timeout_cb = NULL;
+	if (!reply_cb) {
+		LOG_ERR("Undefined reply callback");
+		return -EFAULT;
 	}
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:3041]: This should not be an error. Instead if no callback is provided, use the default one.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 5]
```diff
@@ -3131,3 +3111,13 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
 	return -ENOTSUP;
 #endif
 }
+
+__deprecated
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:3115]:
```suggestion
```
I think this deprecation in C file is unnecessary, just move it into the header.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: include/zephyr/net/lwm2m.h (hunk 1 above)
Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h]: Precise reply_cb as optionnal parameter (NULL)

File: include/zephyr/net/lwm2m.h (hunk 2 above)
Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h:1429]: Done

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 3 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Done

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 4 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Done

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 5 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Done

File: include/zephyr/net/lwm2m.h [hunk 6]
```diff
@@ -1410,7 +1410,23 @@ void lwm2m_rd_client_update(void);
...
+ * @param ctx LwM2M context
+ * @param path_list LwM2M Path string list
+ * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
+ * @param reply_cb Callback triggered with confirmation state or NULL if not used
+ * 
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_engine_send_cb(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
+		     coap_reply_t *reply_cb);
```
Comment (broglep-work) [inline include/zephyr/net/lwm2m.h:1424]: interface design nitpick: using `coap_reply_t` kind of leaks implementation details. Might be worth defining a lwm2m callback for this (like `lwm2m_send_cb_t`)

Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h:1424]: I was thinking about re-encapsulate the callback to be lwm2m related and not coap's.
Do you thing we shall define specific lwm2m code (in callback success/fail) or use the COAP(CODE_CHANGED, etc...) one ,so we still need to import coap.h in that last case

Comment (broglep-work) [reply include/zephyr/net/lwm2m.h:1424]: I have no clear idea yet what would be best. There are success, failure and timeout cases.
Only in failure case you might want to inspect the coap response and would like to know the status code (and maybe even more like options). You made the argument in https://github.com/zephyrproject-rtos/zephyr/issues/52328 that timeout essential means failure for send, which I can also subscribe to. So one option would be to really boil it down to just a boolean, success or failure. 
On the other hand I can imagine use cases were one would like to know if is a hard failure (non-retryable, e.g. rejection by server) or soft failure (retryable, e.g. a timeout). So a result code (int or enum) could come in handy (e.g. success, failure, timeout). And if one really needs it, there could be some additional codes that represent certain COAP response codes. An alternative to bloating the result code with coap specific could be to pass in a context to the callback, that gives access to the coap_packet if there is one available.(that context could even contain a reference the message itself that could no be sent, for convenience of caller)

Comment (SeppoTakalo) [reply include/zephyr/net/lwm2m.h:1424]: I think I'm also leaning towards having a separate callback type `lwm2m_send_cb_t`

It could have just one enum as a parameter. I think we can minimize the cases to what you suggested: OK, rejected, timeout.

And I think this must be first handled inside engine, after all, the timeout is something that the LwM2M engine must react to.

CoAP return code ACCEPTED would be translated to OK/ACCEPTED/SUCCESS. Any other CoAP code would be translated as rejected by the server.
Timeout should cause LwM2M engine to drop into trying to re-register with the server. But timeout should be notified into this callback as well.

Oh.. and another thing, I wan't to get rid of those `lwm2m_engine_something` prefixes, so this would be:

```c
typedef void (*lwm2m_send_cb_t)(enum lwm2m_send_status status);

int lwm2m_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
		    lwm2m_send_cb_t reply_cb);
```

Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h]: Initial callback are unchanged:
`msg->reply_cb = do_send_reply_cb;
msg->message_timeout_cb = do_send_timeout_cb;`

In case of timeout or COAP return success or error, lwm2m_send_cb_t callback is called with the corresponding status
`enum lwm2m_send_status {
	LWM2M_SEND_STATUS_SUCCESS,
	LWM2M_SEND_STATUS_FAILURE,
	LWM2M_SEND_STATUS_TIMEOUT,
};`

In case of multiple send pending, callback is stored in send_status_cb and struct lwm2m_message is stored in void *user_data of struct coap_reply to be retrieve the associated callback.
```
if (reply_cb) {
		msg->reply->user_data = msg;
		msg->send_status_cb = (lwm2m_send_cb_t) reply_cb;
	}
```

**Review by SeppoTakalo — DISMISSED** [review]

**Comment by SeppoTakalo** [comment]
Looks like you need to refactor the existing code to use the new API.
Twister builds are failing because of deprecated code.
See the test results.

**Review by jheiskan81 — DISMISSED** [review]
Looks good for me.

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 7]
```diff
@@ -54,34 +53,27 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
...
-	bool confirmable = true;
 	int ignore_cnt = 1; /* Subcmd + arguments preceding the path list */
 
 	if (!ctx) {
 		shell_error(sh, "no lwm2m context yet\n");
 		return -ENOEXEC;
 	}
 
-	if (argc < 2) {
+	if (argc < 1) {
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:63]: Actually, this is wrong.
argc is always at least one. The `argv[0]` is the `"send"` and first arquement is going to be in `argv[1]`

So don't change this line.

Also, now that there is no optional parameters, you can drop the `ignore_cnt` variable. Previously it was used to skip `-n` from parameter list and assume that all the remaining parameters are just paths. Now we assume that all that follows are paths.

**Comment by RomainPelletant** [comment]
add : when network is disconnected, timeout callback shall be called before reseting messages

**Comment by jheiskan81** [comment]
Looks good please rebase and fix conflict's.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 8]
```diff
@@ -1352,6 +1352,7 @@ int lwm2m_rd_client_stop(struct lwm2m_ctx *client_ctx,
 
 	client.ctx->event_cb = event_cb;
 	rd_client_message_free();
+	lwm2m_engine_context_close(client_ctx);
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1355]: What is this one?
I think this is completely unrelated to this PR

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]
Looking better.

Still the shell command has not been fixed.
Also, there was one unrelated change.

Implementation itself looks OK

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 8 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1355]: @SeppoTakalo thanks for your comment.
After some tests, it appears that if there is a network error or other, timeout callback shall be called on client stop to release send ACKs.
Do you think it shall be done on another PR?

Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c:1355]: Yes, at least create a issue ticket that describes the problem.
I think this is related to issue discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/53476#issuecomment-1400491225

So don't blindly fix, because it might fix only one of the corner cases or might introduce side effects. So at least a separate PR is required, but I would prefer reporting the issue first, so we have a record what is going to be fixed.

**Comment by SeppoTakalo** [comment]
There seem to be conflicts so please rebase on top of main and fix those.

Then fix the shell, like I asked in the first review and drop the one-line rd_client_stop() fix, and I think this is then ready to go in.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c (hunk 8 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_rd_client.c]: Has been removed : I will open a new issue

**Comment by RomainPelletant** [comment]
@broglep-work @SeppoTakalo @jheiskan81 Requested changes has been done

**Comment by SeppoTakalo** [comment]
We still need the `lwm2m_send()` because it is used inside nRF Connect SDK, so this would break the next upmerge.

Also, I spotted faults in the shell handling.
Here is my proposal

```
❯ git diff
  diff --git a/include/zephyr/net/lwm2m.h b/include/zephyr/net/lwm2m.h
  index d866fab9bf..63483f4bc2 100644
  --- a/include/zephyr/net/lwm2m.h
  +++ b/include/zephyr/net/lwm2m.h
  @@ -2104,6 +2104,20 @@ enum lwm2m_send_status {
    */
   typedef void (*lwm2m_send_cb_t)(enum lwm2m_send_status status);
   
  +/** 
  + * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
  + *
  + * @param ctx LwM2M context
  + * @param path_list LwM2M path struct list
  + * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
  + * @param confirmation_request Set to true to use CoAP CON request which is the only allowed.
  + * 
  + * @return 0 for success or negative in case of error.
  + *
  + */
  +int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
  +                         uint8_t path_list_size, bool confirmation_request);
  +
   /** 
    * @brief LwM2M SEND operation to given path list asynchronously with confirmation callback
    *
  diff --git a/subsys/net/lib/lwm2m/lwm2m_message_handling.c b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  index c27630eb8a..6c4b6f7884 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  @@ -2949,8 +2949,18 @@ static bool init_next_pending_timeseries_data(struct lwm2m_cache_read_info *cach
   }
   #endif
   
  +int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
  +                 uint8_t path_list_size, bool confirmation_request)
  +{
  +       if (!confirmation_request) {
  +               return -EINVAL;
  +       }
  +
  +       return lwm2m_send_cb(ctx, path_list, path_list_size, NULL);
  +}
  +
   int lwm2m_send_cb(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
  -                        uint8_t path_list_size, lwm2m_send_cb_t *reply_cb)
  +                 uint8_t path_list_size, lwm2m_send_cb_t *reply_cb)
   {
   #if defined(CONFIG_LWM2M_SERVER_OBJECT_VERSION_1_1)
          struct lwm2m_message *msg;
  diff --git a/subsys/net/lib/lwm2m/lwm2m_shell.c b/subsys/net/lib/lwm2m/lwm2m_shell.c
  index 7b2090f64d..cbfa154645 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_shell.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_shell.c
  @@ -59,7 +59,6 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
          int ret = 0;
          struct lwm2m_ctx *ctx = lwm2m_rd_client_ctx();
          int path_cnt = argc - 1;
  -       int ignore_cnt = 1; /* Subcmd + arguments preceding the path list */
          struct lwm2m_obj_path lwm2m_path_list[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
   
          if (!ctx) {
  @@ -68,12 +67,6 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
          }
   
          if (argc < 2) {
  -               shell_error(sh, "no arguments or path(s)\n");
  -               shell_help(sh);
  -               return -EINVAL;
  -       }
  -
  -       if (argc == 0) {
                  shell_error(sh, "no path(s)\n");
                  shell_help(sh);
                  return -EINVAL;
  @@ -83,18 +76,19 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
                  return -E2BIG;
          }
   
  -       for (int i = ignore_cnt; i < path_cnt; i++) {
  +       for (int i = 0; i < path_cnt; i++) {
                  /* translate path -> path_obj */
  -               ret = lwm2m_string_to_path(argv[i], &lwm2m_path_list[i], '/');
  +               const char *p = argv[1 + i];
  +               ret = lwm2m_string_to_path(p, &lwm2m_path_list[i], '/');
                  if (ret < 0) {
                          return ret;
                  }
          }
   
  -       ret = lwm2m_send_cb(ctx, lwm2m_path_list, path_cnt, NULL);
  +       ret = lwm2m_send(ctx, lwm2m_path_list, path_cnt, true);
   
          if (ret < 0) {
  -               shell_error(sh, "can't do send operation, request failed\n");
  +               shell_error(sh, "can't do send operation, request failed (%d)\n", ret);
                  return -ENOEXEC;
          }
          return 0;
```

File: include/zephyr/net/lwm2m.h [hunk 9]
```diff
@@ -2086,35 +2086,54 @@ void lwm2m_rd_client_update(void);
...
- * @param confirmation_request True request confirmation for operation.
+ * @param reply_cb Callback triggered with confirmation state or NULL if not used
  * 
  * @return 0 for success or negative in case of error.
  *
  */
-int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
-		      bool confirmation_request);
+int lwm2m_send_cb(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
+			  uint8_t path_list_size, lwm2m_send_cb_t *reply_cb);
```
[the remark this answers is not in the document]
Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h]: I was blind, thanks for that check

**Review by SeppoTakalo — DISMISSED** [review]

**Review by jheiskan81 — APPROVED** [review]
Looks good for me.

**Review by jheiskan81 — DISMISSED** [review]
Looks good for me.

**Comment by huelsenfruchtzwerg** [comment]
Just thinking out loud, but wouldn't a user_ctx void * or similar be useful/almost necessary here?
Considering sth like (pseudo-code-is)
`
for (int i = 0; i < 10; ++i) {
    send(data[i], send_cb);
}
[...]
send_cb(status) {
   // now which of my ten messages has been delivered/timed out?
}
`
I suppose you could register ten different cbs in this situation, but that'd be very unwieldy (and impossible once "i" is something dynamic).

**Comment by RomainPelletant** [comment]
@huelsenfruchtzwerg Good point.
From my point of view, lwm2m send method requires SENML_CBOR which is implementing composites reads/write and reduce the needs for several sends.
If I need multiple send, maybe should I only send one message with several objects/data.
Do you think the same way?

**Comment by huelsenfruchtzwerg** [comment]
@RomainPelletant That's also a fair point and something I also considered.
The Problem is when the code calling the send is "dynamic". Simplest (albeit admittedly somewhat constructed) example I can think of is the shell function cmd_send. Currently this doesn't use the send_status_cb, but assuming it did, how would you differentiate between multiple invocations with different arguments in the cb?

If given the choice, I think I'd rather have a single user defined send_cb for all messages and a void* user_ctx for each message. The user could still use the *user_ctx pointer for individual cbs if he so desires, or he could use it for something else.

**Comment by RomainPelletant** [comment]
@huelsenfruchtzwerg Indeed for the shell it could be a critical problem for multiple send.
Currently, we don't receive any confirmation (from user space) and this pull request at least carry out that need.
I think we shall go further and implement a way to register send callback for multilple sends (as you described) : could we consider opening a new issue or that PR shall implement it ASAP?

**Comment by huelsenfruchtzwerg** [comment]
Depends on what approach you/we want to take. Just adding an additional void* ctx parameter to the callback (and the msg struct), would be simple enough to do as part of this pr imho, but would increase the msg structs size by one pointer.
The other approach I outlined above (singular send_status_cb with a void *ctx paramer per send invocation instead of a cb per invocation) would be a bigger change.
I'd defer to @SeppoTakalo and @rlubos, I'm indifferent

File: include/zephyr/net/lwm2m.h [hunk 10]
```diff
@@ -2086,35 +2086,68 @@ void lwm2m_rd_client_update(void);
...
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
  * @param confirmation_request True request confirmation for operation.
  * 
  * @return 0 for success or negative in case of error.
  *
  */
-int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
-	       uint8_t path_list_size, bool confirmation_request);
+__deprecated
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t path_list_size,
```
Comment (rlubos) [inline include/zephyr/net/lwm2m.h:2149]: Why were `lwm2m_engine_send` and `lwm2m_send` reordered? It's an unnecessary change and the diff looks messy because of that.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 11]
```diff
@@ -2879,24 +2879,37 @@ static int do_send_reply_cb(const struct coap_packet *response, struct coap_repl
 			    const struct sockaddr *from)
 {
 	uint8_t code;
+	struct lwm2m_message *msg = NULL;
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:2882]: `msg` could be properly initialized rightaway?

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 12]
```diff
@@ -2937,7 +2950,17 @@ static bool init_next_pending_timeseries_data(struct lwm2m_cache_read_info *cach
 #endif
 
 int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
-	       uint8_t path_list_size, bool confirmation_request)
+			 uint8_t path_list_size, bool confirmation_request)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:2953]: Please don't change proper alignments.

File: include/zephyr/net/lwm2m.h [hunk 13]
```diff
@@ -2086,35 +2086,68 @@ void lwm2m_rd_client_update(void);
...
- * @param path_list LwM2M Path string list
+ * @param path_list LwM2M path struct list
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
- * @param confirmation_request True request confirmation for operation.
+ * @param confirmation_request Set to true to use CoAP CON request which is the only allowed.
+ *
+ * @return 0 for success or negative in case of error.
+ *
+ */
+int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
```
Comment (rlubos) [inline include/zephyr/net/lwm2m.h:2118]: Shouldn't this one be deprecated as well now?

File: subsys/net/lib/lwm2m/lwm2m_shell.c [hunk 14]
```diff
@@ -91,18 +76,20 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
...
+		const char *p = argv[1 + i];
+
+		ret = lwm2m_string_to_path(p, &lwm2m_path_list[i], '/');
 		if (ret < 0) {
 			return ret;
 		}
 	}
 
-	ret = lwm2m_send(ctx, lwm2m_path_list, path_cnt, confirmable);
+	ret = lwm2m_send(ctx, lwm2m_path_list, path_cnt, true);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_shell.c:89]: Shouldn't this use `lwm2m_send_cb()` now?

**Review by rlubos — COMMENTED** [review]
Looks like this needs some post-rebase cleanup.

As for the user ctx for send, I'm not really a fan of extending lwm2m_message structure with non-generic stuff (same would apply to the send_cb). But given the fact, there's no really a better place for now to put it, I guess we could live with that for the time being. @SeppoTakalo 

Ideallly, LwM2M engine could allocate a separate lwm2m_send context, encapsulating lwm2m_message, and extending it with send-specific stuff. This is not very feasible though with preallocated buffers, as we'd need a separate pool just for send contexts. Perhaps, instead of using buffer pools, we could switch at some point to use internal heap for message allocation, then such approach would make more sense. But it's rather tale of the future (if ever).

File: include/zephyr/net/lwm2m.h (hunk 10 above)
Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h]: Modifications reverted

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 11 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Done

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 12 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Done

File: include/zephyr/net/lwm2m.h (hunk 13 above)
Comment (RomainPelletant) [reply include/zephyr/net/lwm2m.h]: You are right : deprecated added

File: subsys/net/lib/lwm2m/lwm2m_shell.c (hunk 14 above)
Comment (RomainPelletant) [reply subsys/net/lib/lwm2m/lwm2m_shell.c:89]: Done

File: include/zephyr/net/lwm2m.h [hunk 15]
```diff
@@ -2106,17 +2124,33 @@ int lwm2m_engine_send(struct lwm2m_ctx *ctx, char const *path_list[], uint8_t pa
...
  * @param ctx LwM2M context
  * @param path_list LwM2M path struct list
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
  * @param confirmation_request True request confirmation for operation.
- * 
+ *
  * @return 0 for success or negative in case of error.
  *
  */
 int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
```
Comment (rlubos) [inline include/zephyr/net/lwm2m.h:2137]: This should use `__deprecated`  macro before the declaration.

File: include/zephyr/net/lwm2m.h [hunk 16]
```diff
@@ -2087,16 +2087,34 @@ void lwm2m_rd_client_update(void);
...
  *
  * @param ctx LwM2M context
  * @param path_list LwM2M Path string list
  * @param path_list_size Length of path list. Max size is CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE
  * @param confirmation_request True request confirmation for operation.
- * 
+ *
  * @return 0 for success or negative in case of error.
  *
  */
```
Comment (rlubos) [inline include/zephyr/net/lwm2m.h:2120]: Same here, `__deprecated` is missing, must've been an overlook.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 17]
```diff
@@ -2935,8 +2947,8 @@ static bool init_next_pending_timeseries_data(struct lwm2m_cache_read_info *cach
 }
 #endif
 
-int lwm2m_send(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
-	       uint8_t path_list_size, bool confirmation_request)
+int lwm2m_send_cb(struct lwm2m_ctx *ctx, const struct lwm2m_obj_path path_list[],
+	       uint8_t path_list_size, lwm2m_send_cb_t reply_cb)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:2951]: Alignment is now broken here.

File: include/zephyr/net/lwm2m.h (hunk 15 above)
Comment (SeppoTakalo) [reply include/zephyr/net/lwm2m.h:2137]: @rlubos Do we have to deprecate this?
If someone don't want to use `_cb` version. I think both can live in the API.

Comment (rlubos) [reply include/zephyr/net/lwm2m.h:2137]: Well, `confirmation_request` is a deprecated parameter, I'm not really sure we want to maintain an API like this. Isn't it good-enough that `lwm2m_send_cb()` accepts NULL as a callback in case it's not used?

Optionally, I guess we could keep the function, but we'd need to modify it's signature, to drop the unused parameter.

**Comment by carlescufi** [comment]
@RomainPelletant ping

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
